Relocate internal Compass docs and implementation details (#1124)
kyma-incubator/compass: components/director/internal/tenantfetcher/README.md

# Tenant Fetcher

## Overview

Tenant Fetcher fetches information about tenants from external APIs.

## Details

This section describes the API schema that a server must implement to integrate with the Tenant Fetcher.

### Authorization

Tenant Events API should use the OAuth 2.0 client credentials authorization flow.

### Endpoints

> There are three types of supported endpoints that receive different type of events:

Reviewer: ```suggestion
There are three types of supported endpoints that receive different types of events:
```

- Tenant creation endpoint
- Tenant deletion endpoint
- Tenant update endpoint

> Every endpoints must return a specific payload and accept the following query parameters:

Reviewer: ```suggestion
Every endpoint must return a specific payload and accept the following query parameters:
```
Oops, we missed a typo here!

> - **ts** - specifies a timestamp in Unix time, that is the date since events are fetched

Reviewer: ```suggestion
- **ts** - specifies a timestamp in Unix time, that is the date from which events are fetched
```
Or `since` + Present Perfect, but I think that this sound better.

> - **page** - specifies the number of page to be fetched, starting from `1`

Reviewer: ```suggestion
- **page** - specifies the number of the page to be fetched, starting from `1`
```

- **resultsPerPage** - specifies the number of results included on a single page

#### Tenant creation endpoint

> The endpoint should return the following JSON payload:

Reviewer: `Should`? So what happens if it doesn't?
```suggestion
On success, the endpoint returns the following JSON payload:
```
Or just `The endpoint returns the following JSON payload:` if the payload is returned regardless of the operation result.

See the [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/08-style-and-terminology.md#terminology).

```json
{
  "events": [
    {
      "eventData": "{\"$id\":\"837d023b-782d-4a97-9d38-fecab47c296a\",\"$name\":\"Tenant 1\",\"$discriminator\":\"default\"}"
    }
  ],
  "totalResults": 27,
  "totalPages": 1
}
```

The **eventData** field contains an escaped JSON string with the following fields that you can configure using values overrides:
- `$id` - specifies a unique tenant ID
- `$name` - specifies the tenant name
- `$discriminator` - specifies an optional field that can be used to distinguish different types of tenants

#### Tenant deletion endpoint

> The endpoint should return the following JSON payload:

Reviewer: `Should`? So what happens if it doesn't?
```suggestion
On success, the endpoint returns the following JSON payload:
```
Or just `The endpoint returns the following JSON payload:` if the payload is returned regardless of the operation result.

See the [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/08-style-and-terminology.md#terminology).

```json
{
  "events": [
    {
      "eventData": "{\"$id\":\"837d023b-782d-4a97-9d38-fecab47c296a\",\"$name\":\"Tenant 1\"}"
    }
  ],
  "totalResults": 27,
  "totalPages": 1
}
```

The **eventData** field contains an escaped JSON string with the following fields that you can configure using values overrides:
- `$id` - specifies a unique tenant ID
- `$name` - specifies the tenant name

#### Tenant update endpoint

> The endpoint should return the following JSON payload:

Reviewer: `Should`? So what happens if it doesn't?
```suggestion
On success, the endpoint returns the following JSON payload:
```
Or just `The endpoint returns the following JSON payload:` if the payload is returned regardless of the operation result.

See the [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/08-style-and-terminology.md#terminology).

```json
{
  "events": [
    {
      "eventData": "{\"$id\":\"837d023b-782d-4a97-9d38-fecab47c296a\",\"$name\":\"Tenant 1\"}"
    }
  ],
  "totalResults": 27,
  "totalPages": 1
}
```

The **eventData** field contains an escaped JSON string with the following fields that you can configure using values overrides:
- `$id` - specifies a unique tenant ID
- `$name` - specifies the tenant name

## Configuration

Tenant Fetcher binary allows you to override some configuration parameters. You can specify the following environment variables:

| Parameter | Description |  Default value |
|-----------|-------------|---------------|

> | **global.tenantFetcher.enabled** | Parameter that enables the Tenant Fetcher CronJob | `false` |

Reviewer: ```suggestion
| **global.tenantFetcher.enabled** | Parameter enabling the Tenant Fetcher CronJob | `false` |
```
This one is just a loose suggestion. If you think it sounds better, go ahead and apply it. If not, please ignore it. 

I like the verbal sentence better than the nominal sentence here, but 
1. The verbal sentence is lacking `The` in front. 
2. The nominal sentence is more cohesive with the rest of the descriptions in the table.

Author: I also like the verbal sentence better. Sounds more natural to me so I'll stay with it.

| **global.tenantFetcher.providerName** | Name of the tenants provider | `"compass"` |

> | **global.tenantFetcher.schedule** | Parameter that specifies how often Tenant Fetcher fetches information about tenants | `"*/5 * * * *"` |

Reviewer: ```suggestion
| **global.tenantFetcher.schedule** | Parameter specifying how often Tenant Fetcher fetches information about tenants | `"*/5 * * * *"` |
```
This one is just a loose suggestion. If you think it sounds better, go ahead and apply it. If not, please ignore it. 

I like the verbal sentence better than the nominal sentence here, but 
1. The verbal sentence is lacking `The` in front. 
2. The nominal sentence is more cohesive with the rest of the descriptions in the table.

Author: I also like the verbal sentence better. Sounds more natural to me so I'll stay with it.

| **global.tenantFetcher.oauth.client** | OAuth 2.0 client ID | None |
| **global.tenantFetcher.oauth.secret** | OAuth 2.0 client Secret | None |
| **global.tenantFetcher.oauth.tokenURL** | Endpoint for fetching an OAuth 2.0 access token to the Tenant Events API | None |
| **global.tenantFetcher.endpoints.tenantCreated** | Tenant Events API endpoint for fetching created tenants | `"127.0.0.1/events?type=created"` |
| **global.tenantFetcher.endpoints.tenantDeleted** | Tenant Events API endpoint for fetching deleted tenants | `"127.0.0.1/events?type=deleted"` |
| **global.tenantFetcher.endpoints.tenantUpdated** | Tenant Events API endpoint for fetching updated tenants | `"127.0.0.1/events?type=updated"` |
| **global.tenantFetcher.fieldMapping.idField** | Name of the field in the event data payload containing the tenant name | `"id"` |

> | **global.tenantFetcher.fieldMapping.nameField** | Name of the field in the event data payload containing the tenant ID | `"name"` |

Reviewer: Aren't these two mixed up..? Look at the parameter names, their descriptions, and default values. I think the descriptions are the other way round.
